l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -559,82 +487,59 @@ contract L1SharedBridge is IL1SharedBridge, ReentrancyGuard, Ownable2StepUpgrade
         require(!_isEraLegacyDeposit(_chainId, _l2BatchNumber, _l2TxNumberInBatch), "ShB: legacy cFD");
         {
             bytes32 dataHash = depositHappened[_chainId][_l2TxHash];
-            address l1Token = nativeTokenVault.tokenAddress(_assetId);
-            (uint256 amount, address prevMsgSender) = abi.decode(_transferData, (uint256, address));
-            bytes32 txDataHash = keccak256(abi.encode(prevMsgSender, l1Token, amount));
-            require(dataHash == txDataHash, "ShB: d.it not hap");
+            // Determine if the given dataHash matches the calculated legacy transaction hash.
+            bool isLegacyTxDataHash = _isLegacyTxDataHash(_depositSender, _assetId, _assetData, dataHash);
+            // If the dataHash matches the legacy transaction hash, skip the next step.
+            // Otherwise, perform the check using the new transaction data hash encoding.
+            if (!isLegacyTxDataHash) {
+                bytes32 txDataHash = this.encodeTxDataHash(false, _depositSender, _assetId, _assetData);
```

Comment:
nit: we could create an internal method for situations like this. This is a free way to save up on gas to not do additional calls/abi encoding/decoding etc.